rio/src/RIO/Vector.hs

```diff
@@ -257,3 +257,20 @@ module RIO.Vector
   ) where
 
 import qualified Data.Vector.Generic
+import Data.Vector.Generic (Vector)
+import Control.Monad (guard)
+
+-- | /O(1)/ Yield a slice of the vector without copying it. If the vector
+-- cannot satisfy the specificed slice 'Nothing' is returned.
+--
+-- @since 0.1.13.0
+sliceMaybe :: Vector v a
+  => Int  -- ^ @i@ starting index
+  -> Int  -- ^ @n@ length
+  -> v a
+  -> Maybe (v a)
+sliceMaybe i n v = do
+  guard $ i >= 0
+  guard $ n >= 0
+  guard $ Data.Vector.Generic.length v >= i + n
```

Comment:
I'm fairly certain this is subject to an integer overflow error: if `i` and `n` are both very large, they'll be positive, their sum will overflow and can be negative, and then it will be less than `length v`.

Reply:
Good point. What is the canonical way to check for this?
1.  `i + n >= i`
2.  `n <= maxBound - i`
3.  …?

Reply:
I'm fairly certain you can replace your current check with this:

```haskell
guard $ Data.Vector.Generic.length v - n >= i
```

We know that `length v`, `n` and `i` are all non-negative, and there are more negative `Int`s than positive (due to design of 2's complement). Therefore, `length v - n` is guaranteed not to overflow. That means that, if `n` is too large, `length v - n` is guaranteed to be negative, which is guaranteed to fail the test.

Reply:
Agreed. Thanks